shared/src/balancer_event_handler.rs

```diff
@@ -3,195 +3,183 @@ use crate::{
+        Ok(vault
+            .methods()
+            .get_pool_tokens(Bytes(pool_id.0))
+            .call()
+            .await?
+            .0)
+    }
+
+    async fn get_pool_id(&self, pool_address: H160) -> Result<H256> {
+        let pool_contract = BalancerV2WeightedPool::at(self, pool_address);
+        Ok(H256::from(
+            pool_contract.methods().get_pool_id().call().await?.0,
+        ))
+    }
+
+    async fn get_pool_data(&self, pool_address: H160) -> Result<WeightedPoolData> {
+        let pool_id = self.get_pool_id(pool_address).await?;
+        let tokens = self.get_pool_tokens(pool_id).await?;
+        let weights = self.get_normalized_weights(pool_address).await?;
+        Ok(WeightedPoolData {
+            pool_id,
+            tokens,
+            weights,
+        })
     }
 }
 
 /// The BalancerPool struct represents in-memory storage of all deployed Balancer Pools
-#[derive(Debug, Default)]
+#[derive(Derivative)]
+#[derivative(Debug)]
 pub struct BalancerPools {
     /// Used for O(1) access to all pool_ids for a given token
     pools_by_token: HashMap<H160, HashSet<H256>>,
     /// WeightedPool data for a given PoolId
     pools: HashMap<H256, WeightedPool>,
-    /// Temporary storage for WeightedPools containing insufficient constructor data
-    pending_pools: HashMap<H256, WeightedPoolBuilder>,
-}
-
-/// There are three specialization settings for Pools,
-/// which allow for cheaper swaps at the cost of reduced functionality:
-#[derive(Debug, Clone, Copy, Eq, Hash, PartialEq)]
-#[repr(u8)]
-pub enum PoolSpecialization {
-    /// no specialization, suited for all Pools. IGeneralPool is used for swap request callbacks,
-    /// passing the balance of all tokens in the Pool. These Pools have the largest swap costs
-    /// (because of the extra storage reads), which increase with the number of registered tokens.
-    General = 0,
-    /// IMinimalSwapInfoPool is used instead of IGeneralPool, which saves gas by only passing the
-    /// balance of the two tokens involved in the swap. This is suitable for some pricing algorithms,
-    /// like the weighted constant product one popularized by Balancer V1. Swap costs are
-    /// smaller compared to general Pools, and are independent of the number of registered tokens.
-    MinimalSwapInfo = 1,
-    /// only allows two tokens to be registered. This achieves the lowest possible swap gas cost.
-    /// Like minimal swap info Pools, these are called via IMinimalSwapInfoPool.
-    TwoToken = 2,
-}
-
-impl PoolSpecialization {
-    fn new(specialization: u8) -> Result<Self> {
-        match specialization {
-            0 => Ok(Self::General),
-            1 => Ok(Self::MinimalSwapInfo),
-            2 => Ok(Self::TwoToken),
-            t => Err(anyhow!("Invalid PoolSpecialization value {} (> 2)", t)),
-        }
-    }
+    /// Temporary storage for WeightedPools having errored on data fetching
+    pending_pools: HashSet<WeightedPoolBuilder>,
```

Comment:
I'm not sure I understand the use of a `HashSet` here - since we don't expect two created events to come in for the same pool right? Also is this really needed now that we only need a single event in order to register a weighted pool?

Reply:
I have answered this above but will rework pending pools into a vector instead. The reason for this as a set instead of a map is that we no longer have pool_id at event emission time. So I turned it from HashMap to HashSet. The only other option for use here is to use a Vector if that is something you would prefer. It is assumed that we would never get the same event twice since I don't believe that could ever happen.

Reply:
I don't think we even need this in the first place - to me it looks like these are added to the `pending_pools` member temporarily in `insert_events` only to be upgraded at the end. I think we can just construct a list of `WeightedPool`s from a list of `PoolCreated` events with a block number right?

Reply:
As I have mentioned above, this try_upgrade logic with the pool builder has been left here so that we can make batch_calls to the EVM for pool token data.